virtual_ecosystem/models/animal/animal_model.py

```diff
@@ -769,131 +774,231 @@ def remove_dead_cohort_community(self) -> None:
             self.remove_dead_cohort(cohort)
 
     def birth(self, parent_cohort: AnimalCohort) -> None:
-        """Produce a new AnimalCohort through reproduction.
+        """Produce offspring for a parent cohort using helper methods.
 
-        A cohort can only reproduce if it has an excess of reproductive mass above a
-        certain threshold. The offspring will be an identical cohort of adults
-        with age 0 and a mass determined by the functional group's birth mass.
+        This orchestrates the reproduction process, including:
+        - Calculating total available reproductive mass.
+        - Determining number of offspring.
+        - Creating offspring and adding them to the population.
+        - Updating parent mass after reproduction.
+        - Removing semelparous parents if applicable.
 
         Args:
-            parent_cohort: The AnimalCohort instance which is producing a new cohort.
+            parent_cohort: The parent cohort giving birth.
         """
-        # Handle semelparous reproduction (where parents die after reproduction)
-        non_reproductive_mass_loss_c = non_reproductive_mass_loss_n = (
-            non_reproductive_mass_loss_p
-        ) = 0.0
-
-        if parent_cohort.functional_group.reproductive_type == "semelparous":
-            loss_c = (
-                parent_cohort.mass_cnp.carbon
-                * parent_cohort.constants.semelparity_mass_loss
-            )
-            loss_n = (
-                parent_cohort.mass_cnp.nitrogen
-                * parent_cohort.constants.semelparity_mass_loss
-            )
-            loss_p = (
-                parent_cohort.mass_cnp.phosphorus
-                * parent_cohort.constants.semelparity_mass_loss
-            )
+        reproductive_mass = self.calculate_total_reproductive_mass(parent_cohort)
+        number_offspring = self.calculate_offspring_count(
+            parent_cohort, reproductive_mass
+        )
 
-            # Cap loss to available mass
-            non_reproductive_mass_loss_c = min(loss_c, parent_cohort.mass_cnp.carbon)
-            non_reproductive_mass_loss_n = min(loss_n, parent_cohort.mass_cnp.nitrogen)
-            non_reproductive_mass_loss_p = min(
-                loss_p, parent_cohort.mass_cnp.phosphorus
-            )
+        if number_offspring <= 0:
```

Comment:
Are there situations in which `calculate_offspring_count` can return numbers <0? That seems a bit weird. Or is it just defensive programming? If it can't ever be <0, I'd just do this for clarity:

```suggestion
        if number_offspring == 0:
```